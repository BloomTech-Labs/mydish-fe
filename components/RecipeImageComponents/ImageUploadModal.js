import React from 'react';
import { View, TouchableOpacity, Alert, Image, Text } from 'react-native';
import Modal from 'react-native-modal';
import * as ImagePicker from 'expo-image-picker';
import * as Permissions from 'expo-permissions';
import styles from '../../styles/recipeImageStyles';
import { useDispatch } from 'react-redux';
import { editImage } from '../../store/singleRecipe/singleRecipeActions';
import {
  generateIngredients,
  generateInstructions,
} from '../../store/generate/generateRecipeAction';
import camera from '../../assets/camera_red.png';
import gallery from '../../assets/image_red.png';

function ImageUploadModal({
  visible,
  setVisible,
  setRecipe,
  setIngredients,
  setInstructions,
  parent,
}) {
  const dispatch = useDispatch();
  const take = 'take';
  const choose = 'choose'; // Pass take or choose as argument to getImage()

  const verifyPermissions = async () => {
    const result = await Permissions.askAsync(
      Permissions.CAMERA,
      Permissions.CAMERA_ROLL
    );
    if (result.status !== 'granted') {
      Alert.alert(
        'Insufficient Permissions',
        'You must grant permissions to upload a photo.',
        [{ text: 'Okay' }]
      );
      return false;
    } else {
      return true;
    }
  };
  // @param {string} method
  //(value of "take" will launch camera, "choose" will open image gallery)
  const getImage = async (method) => {
    const hasPermission = await verifyPermissions();
    if (!hasPermission) return;

    let imgConfig = {};

    parent === 'generateIngredients' || parent === 'generateInstructions'
      ? (imgConfig = {
          base64: true,
          quality: 0.5,
          allowsEditing: true,
        })
      : (imgConfig = {
          allowsEditing: true,
          aspect: [16, 9],
          quality: 0.5,
        });

    let img = '';

    if (method === take) {
      img = await ImagePicker.launchCameraAsync(imgConfig);
    } else if (method === choose) {
      img = await ImagePicker.launchImageLibraryAsync(imgConfig);
    }

    if (img && !img.cancelled) {
      if (parent === 'create') {
        setRecipe((oldRecipe) => ({
          ...oldRecipe,
          img: img.uri,
        }));
      } else if (parent === 'editRecipe') {
        dispatch(editImage(img.uri));
      } else if (parent === 'generateIngredients') {
        dispatch(generateIngredients(img.base64));
        setIngredients(false);
      } else if (parent === 'generateInstructions') {
        dispatch(generateInstructions(img.base64));
        setInstructions(false);
      }
    }
    setVisible(false);
  };

  return (
    <Modal
      animationIn="fadeIn"
      animationOut="fadeOut"
      animationInTiming={400}
      animationOutTiming={400}
      hideModalContentWhileAnimating={true}
      isVisible={visible}
      backdropOpacity={0.4}
      onBackdropPress={() => setVisible(false)}
    >
      <View style={styles.uploadModal}>
        <View style={styles.iconsContainer}>
          <TouchableOpacity
            style={styles.singleIconContainer}
            testID="cameraIcon"
            onPress={() => getImage(take)}
          >
            <Image
              style={{
                ...styles.iconLarge,
                ...styles.iconLargeCamera,
              }}
              source={camera}
            />
            <Text style={styles.text}>Camera</Text>
          </TouchableOpacity>
          <TouchableOpacity
            testID="imageIcon"
            style={styles.singleIconContainer}
            onPress={() => getImage(choose)}
          >
            <Image style={styles.iconLarge} source={gallery} />
            <Text style={styles.text}>Gallery</Text>
          </TouchableOpacity>
        </View>
      </View>
    </Modal>
  );
}

export default ImageUploadModal;
