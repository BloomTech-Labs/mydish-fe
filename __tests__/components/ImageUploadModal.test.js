import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import ImageUploadModal from '../../components/RecipeImageComponents/ImageUploadModal';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

describe('<ImageUploadModal />', () => {
  const store = createStore(() => {});
  const renderTree = (
    <Provider store={store}>
      <ImageUploadModal />
    </Provider>
  );
  const tree = renderer.create(renderTree).toJSON();
  it('matches snapshot', () => {
    expect(tree).toMatchSnapshot();
  });
});
