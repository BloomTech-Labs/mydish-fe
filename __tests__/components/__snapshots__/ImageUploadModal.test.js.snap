// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageUploadModal /> matches snapshot 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={true}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  visible={false}
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    hideModalContentWhileAnimating={true}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    pointerEvents="box-none"
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "margin": 37.5,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F1F1F1",
          "borderRadius": 10,
          "height": 122,
          "justifyContent": "center",
          "marginHorizontal": 59,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-around",
            "width": "90%",
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="cameraIcon"
        >
          <Image
            source={1}
            style={
              Object {
                "height": 42,
                "marginBottom": 8,
                "width": 49,
              }
            }
          />
          <Text
            style={
              Object {
                "fontFamily": "nunito sans bold",
                "fontSize": 12,
                "paddingVertical": 2,
              }
            }
          >
            Camera
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="imageIcon"
        >
          <Image
            source={1}
            style={
              Object {
                "height": 49,
                "marginBottom": 8,
                "width": 49,
              }
            }
          />
          <Text
            style={
              Object {
                "fontFamily": "nunito sans bold",
                "fontSize": 12,
                "paddingVertical": 2,
              }
            }
          >
            Gallery
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
